p2p/src/simulated/network.rs

```diff
@@ -284,27 +298,90 @@ impl<E: RNetwork + Spawner + Rng + Clock + Metrics, P: PublicKey> Network<E, P>
-            // Apply link settings
-            let delay = link.sampler.sample(&mut self.context);
+            // Sample latency and get current time
+            let latency = Duration::from_millis(link.sampler.sample(&mut self.context) as u64);
+            let now = self.context.current();
+
+            // Get peer bandwidth info and calculate effective bandwidth
+            let (sender_egress_bps, sender_egress_available_at) = {
+                let sender = self.peers.get(&origin).unwrap();
+                (sender.egress_bps, sender.egress_available_at)
+            };
+
+            let (receiver_ingress_bps, receiver_ingress_available_at) = {
+                let receiver = self.peers.get(&recipient).unwrap();
+                (receiver.ingress_bps, receiver.ingress_available_at)
+            };
+
+            let effective_bps = match (sender_egress_bps, receiver_ingress_bps) {
+                (Some(egress), Some(ingress)) => Some(egress.min(ingress)),
+                (Some(egress), None) => Some(egress),
+                (None, Some(ingress)) => Some(ingress),
+                (None, None) => None,
+            };
+
+            // Calculate transmission timing
+            let (send_complete_at, mut receive_complete_at) = if let Some(bps) = effective_bps {
+                let tx_duration =
+                    Duration::from_millis((message.len() as f64 / bps as f64 * 1000.0) as u64);
+
+                // Sender can start when free, receiver must be free when first bit arrives
+                let sender_ready_at = sender_egress_available_at.max(now);
+                let receiver_ready_at = receiver_ingress_available_at
+                    .checked_sub(latency)
+                    .unwrap_or(SystemTime::UNIX_EPOCH);
+
+                let send_complete_at = sender_ready_at.max(receiver_ready_at) + tx_duration;
+                let receive_complete_at = send_complete_at + latency;
+
+                (send_complete_at, receive_complete_at)
+            } else {
+                // No bandwidth limits
+                (now, now + latency)
+            };
+
+            // Apply ordering constraint: messages must arrive after previous message on this link
+            // (we add a 1 microsecond epsilon to ensure strict ordering)
+            receive_complete_at =
+                receive_complete_at.max(link.last_arrival_at + Duration::from_micros(1));
+
+            // Update the link's last arrival time
+            link.last_arrival_at = receive_complete_at;
+
+            // Update availability times for peers with bandwidth limits
+            if sender_egress_bps.is_some() {
+                self.peers.get_mut(&origin).unwrap().egress_available_at = send_complete_at;
+            }
+
+            if receiver_ingress_bps.is_some() {
+                self.peers.get_mut(&recipient).unwrap().ingress_available_at = receive_complete_at;
+            }
+
             let should_deliver = self.context.gen_bool(link.success_rate);
```

Comment:
I guess it would be "most correct" if we determined `should_deliver` prior to updating the receiver's limit?

I suppose we weren't very clear on the fault model for not delivered but in my head, I've been thinking of it as sender uses their egress but never gets to recipient.

Reply:
I think that's fair. Will update.